doc/feature/cppPlugins.md

```diff
@@ -0,0 +1,159 @@
+
+    class PluginRegistry
+    {
+    public:
+        typedef std::unique_ptr< Compressor > CompressorPtr;
+        CompressorPtr choose();
+        CompressorPtr create( const std::string& name );
+    };
+
+    class Compressor
+    {
+    public:
+        typedef lunchbox::Bufferb Result;
+        typedef std::vector< Result > Results;
+
+        virtual const Results& compress( const void* const data,
+                                         const size_t size ) = 0;
+        virtual const Result& decompress( const Results& input,
+                                          const size_t size ) = 0;
+    protected:
+        Results compressed;
+        Result uncompressed;
+    };
+
+
+## Examples
+
+    CompressorPtr PluginRegistry::choose()
+    {
+        float rating = powf( 1.0f, .3f );
+        Plugin* candidate = nullptr;
+
+        const Plugins& plugins = PluginFactory::getInstance().getPlugins();
+        for( const Plugin* plugin : plugins )
+        {
+            for( const PluginInfo& info : plugin->getInfo( ))
+            {
+                float newRating = powf( info.speed, .3f ) / info.ratio;
+                if( newRating > rating )
+                {
+                    candidate = plugin;
+                    rating = newRating;
+                }
+            }
+        }
+        if( candidate )
+            return candidate.construct()
+        return nullptr;
+    }
+
+Compressor implementation:
+
+    #include "zstd/lib/zstd.h"
+
+    namespace pression
+    {
+    namespace plugin
+    {
+    namespace
+    {
+    static const bool LB_UNUSED _initialized = []()
+    {
+        PluginRegistry::registerEngine( { "pression::CompressorZSTD",
+                                         .47f, .25f },
+                                        CompressorZSTD::newInstance );
+        return true;
+    }();
+    }
+
+    const Results& CompressorZSTD::compress( const void* const data,
+                                             const size_t inSize )
+    {
+        if( _results.empty( ))
+            _results.push_back( new pression::plugin::Compressor::Result );
+        _results[0]->reserve( ZSTD_compressBound( inSize ));
+
+        const size_t size = ZSTD_compress( _results[0]->getData(),
+                                           _results[0]->getSize(), inData,
+                                            inSize, 2 /* level */ );
+        _results[0]->setSize( size );
+        return results;
+    }
+
+    const Result& CompressorZSTD::decompress( const Results& input,
+                                              const size_t size );
+    {
+        if( input.empty( ))
+            return;
+
+        _result.resize( size );
+        ZSTD_decompress( _result.getData(), _result.getSize(),
+                         input[0].getData(), input[0].getSize( ));
+    }
+    }
+    }
+
+## Issues
+
+### Issue 1: How do we select compression engines?
+
+_Resolution: Make InitDataT& parameter given to handles() mutable_
```

Comment:
Is this still valid? I thought this was solved differently now. I also don't see this in the example above,